README.md
# Robin

[![Build Status](https://travis-ci.org/nib-health-funds/robin.svg?branch=master)](https://travis-ci.org/nib-health-funds/robin)
[![Dependencies](https://david-dm.org/nib-health-funds/robin.svg)](https://david-dm.org/nib-health-funds/robin)

Batman's very capable side kick - deletes old ECR images.
<center><img src="images/robin.jpg"</img></center>

# Why Robin?

Images stored in ECR incur monthly data storage charges, this means paying to store old images that are no longer in use. Also, AWS ECR has a default limit of 1000 images. Therefore, it is desirable to ensure the ECR repositories are kept clean of unused images.

# What we delete currently:

- Images older than 30 days that do not have tags that contain 'master'

# Usage

1. Authenticate and get AWS credentials via your preferred CLI, you may need to export the environment variables directly

1. Update the environment variables in serverless.yml to match your setup

1. Update the `repoNames` array in `handler.js` to reflect the repositories you want to clean

1. Lint the function

```
$ npm run lint
```

5. Test the function

```
$ npm test
```

6. Deploy the function

```
$ npm run deploy
```

7. Tail cloudwatch logs

```
$ npm run tail-logs
```


# TODO

- Only keep the last 10 master images (justification: we should be using the last images only, last 10 gives us something to rollback to if needed.)
- Add some more documentation to this readme
- Delete all untagged images
- Make repo names configurable via env vars
- Make tagging convention configurable

CHANGELOG.md
### 1.2.0

- Mock slack integration
- Add dry run

### 1.1.2

- Add slack integration

### 1.0.1

- Adds multiple repository support
- No longer sends empty array of images to delete
- Adds linting support
- Removes unused code
- Adds linting pre-push stage
- Adds documentation to the readme

handler.test.js
const assert = require('assert');
const moment = require('moment');
const proxyquire = require('proxyquire');
const sinon = require('sinon');

const deletePromise = {
  promise: () => Promise.resolve({
    imageIds: [{
      imageDigest: '3797470f-bb69-11e6-90e4-19b39eb619f7',
      imageTag: 'some-tag'
    }],
    failures: []
  })
};

const desribeImagesPromise = {
  promise: () => Promise.resolve(
    {
      nextToken: null,
      imageDetails: [{
        registryId: '384553929753',
        repositoryName: 'test-repo',
        imageDigest: '1',
        imageTags: ['1.0.0-master', 'ignore-this'],
        imageSizeInBytes: '1024',
        imagePushedAt: moment()
      },
      {
        registryId: '384553929753',
        repositoryName: 'test-repo',
        imageDigest: '2',
        imageTags: ['1.0.0-master', 'ignore-this'],
        imageSizeInBytes: '1024',
        imagePushedAt: moment().add(-31, 'days')
      },
      {
        registryId: '384553929753',
        repositoryName: 'test-repo',
        imageDigest: '3',
        imageTags: ['1.0.0-other', 'ignore-this'],
        imageSizeInBytes: '1024',
        imagePushedAt: moment()
      },
      {
        registryId: '384553929753',
        repositoryName: 'test-repo',
        imageDigest: '4',
        imageTags: ['1.0.0-other', 'dont-ignore-this'],
        imageSizeInBytes: '1024',
        imagePushedAt: moment().add(-31, 'days')
      }]
    })
};

const sandbox = sinon.sandbox.create();
const deleteStub = sandbox.stub();
const describeImagesStub = sandbox.stub().returns(desribeImagesPromise);
const rpStub = sandbox.stub();
const mocks = {
  'aws-sdk': {
    ECR: function () { // eslint-disable-line
      this.batchDeleteImage = deleteStub;
      this.describeImages = describeImagesStub;
    }
  },
  'request-promise': rpStub
};

describe('cleanupImages', () => {
  const cleanup = proxyquire('./handler.js', mocks);

  beforeEach(() => {
    sandbox.reset();
  });

  it('Should not remove master images', done => {
    deleteStub.returns(deletePromise);
    process.env.DRY_RUN = false;
    cleanup.cleanupImages(null, null, () => {
      assert(describeImagesStub.called);
      assert(deleteStub.calledWith({
        registryId: '441581275790',
        repositoryName: 'test-repo',
        imageIds: [{ imageDigest: '4' }]
      }));
      assert(rpStub.called);
      done();
    });
  });

  it('Should not call delete if dry run', done => {
    deleteStub.returns(deletePromise);
    process.env.DRY_RUN = true;
    cleanup.cleanupImages(null, null, () => {
      assert(describeImagesStub.called);
      assert(deleteStub.notCalled);
      assert(rpStub.called);
      done();
    });
  });
});


handler.js
'use strict';

const AWS = require('aws-sdk');
const moment = require('moment');
const filter = require('lodash.filter');

const registry = '441581275790';
const repoNames = [
  'test-repo'
];

const rp = require('request-promise');

function postToSlack(deleted, failed) {
  const removedText = `${deleted === 'none' ? 0 : deleted.length} ECR images have been removed.`;
  const failedText = `${failed === 'none' ? 0 : failed.length} ECR images encounted problems when trying to remove them.`;
  const options = {
    method: 'POST',
    uri: process.env.SLACK_WEBHOOK,
    body: { text: `${removedText}\n${failedText}` },
    json: true
  };
  return rp(options);
}

module.exports.cleanupImages = (event, context, callback) => {
  console.log('Robin is dealing out some of his own justice...');
  
  const isDryRun = process.env.DRY_RUN === 'true';
  console.log('Robin is running in dry run mode: ', isDryRun);

  const ecrRegion = process.env.ECR_REGION || 'us-east-1';
  console.log('Robin is using ECR Region: ', ecrRegion);

  const ecr = new AWS.ECR({ apiVersion: '2015-09-21', region: ecrRegion });

  const cutOffDate = moment().add(-30, 'd');
  console.log('Using cut off date: ', cutOffDate);

  const promises = repoNames.map(repoName => {
    const params = {
      registryId: registry,
      repositoryName: repoName,
      maxResults: 100
    };

    return ecr.describeImages(params)
      .promise()
      .then(images => { // eslint-disable-line arrow-body-style
        return filter(images.imageDetails, image => (
          // filters out images that are 30 days old and don't contain the master tag
          moment(image.imagePushedAt).isBefore(cutOffDate)
          && !image.imageTags.find(tag => tag.indexOf('master') > -1)
        ));
      })
      .then(toDelete => {
        console.log('Images to delete: ', toDelete);
        if (!toDelete || toDelete.length === 0) return Promise.resolve({ imageIds: 'none', failures: 'none' });
        const convertedToDelete = toDelete.map(image => ({ imageDigest: image.imageDigest }));
        const deleteParams = {
          registryId: registry,
          repositoryName: repoName,
          imageIds: convertedToDelete
        };

        if (isDryRun) {
          return Promise.resolve({ imageIds: [], failures: []});
        }

        return ecr.batchDeleteImage(deleteParams).promise();
      })
      .then(deleteData => {
        console.log('Deleted images: ', deleteData.imageIds);
        console.log('Delete failures: ', deleteData.failures);
        return postToSlack(deleteData.imageIds, deleteData.failures);
      });
  });

  return Promise.all(promises)
    .then(() => {
      callback(null, { message: 'robin executed successfully!', event });
    })
    .catch(err => {
      console.log(err); // an error occurred
      callback(err);
      return;
    });
};
